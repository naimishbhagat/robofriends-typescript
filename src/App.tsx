import React, { useState, useEffect } from "react";

import { RobotType } from "./types";
import CardList from "./components/CardList";
import SearchBox from "./components/SearchBox";
import "./App.css";

function App() {
  const [robots, setRobots] = useState<RobotType[]>([]);
  const [error, setError] = useState<string | boolean>(false);
  const [searchField, setSearchField] = useState<string>("");
  const getRobots = async (): Promise<void> => {
    try {
      const fetchData = await fetch(
        "https://jsonplaceholder.typicode.com/users"
      );

      if (fetchData.ok) {
        const response = await fetchData.json();
        return response;
      } else if (fetchData.status > 200 && fetchData.status < 500) {
        return Promise.reject("Error 404");
      } else {
        return Promise.reject("Some other error: " + fetchData.status);
      }
    } catch (err) {
      return Promise.reject(err.message);
    }
  };
  useEffect(() => {
    let mounted = true;

    getRobots()
      .then((data) => {
        if (mounted) {
          setRobots(data as any);
        }
      })
      .catch((err) => {
        if (mounted) {
          setError(err);
        }
      });
    return () => {
      mounted = false;
    };
  }, []);

  const onSearchChange = (event: React.ChangeEvent<HTMLInputElement>): void => {
    setSearchField(event.target.value);
  };
  if (error) {
    return <div data-testid="error">{error}</div>;
  }
  const filteredRobots =
    robots.length > 0
      ? robots.filter((robot) => {
          return robot.name.toLowerCase().includes(searchField.toLowerCase());
        })
      : [];
  return !robots.length ? (
    <h1 data-testid="loading">Loading</h1>
  ) : (
    <div className="tc">
      <h1 className="f1">Robofriends</h1>
      <SearchBox searchChange={onSearchChange} searchField={searchField} />
      <CardList robots={filteredRobots} />
    </div>
  );
}

export default App;
