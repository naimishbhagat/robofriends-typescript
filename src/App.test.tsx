import React from "react";
import {
  render,
  screen,
  waitForElementToBeRemoved,
} from "@testing-library/react";
import { unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";

import App from "./App";
import userEvent from "@testing-library/user-event";
const stubbedData = [
  {
    id: 1,
    name: "Naimish Bhagat",
    username: "nb",
    email: "nb@april.biz",
  },
  {
    id: 2,
    name: "Mitchell Bhagat",
    username: "mb",
    email: "mb@melissa.tv",
  },
];

function fetchMock() {
  return new Promise((resolve) =>
    resolve({
      ok: true,
      status: 200,
      json: () => Promise.resolve(stubbedData),
    })
  );
}

function rejectFetchMock404() {
  return new Promise((reject) =>
    reject({
      ok: false,
      status: 404,
      json: () => Promise.reject("Not found"),
    })
  );
}

function rejectFetchMock() {
  throw new Error("Unknown Error");
}

let container: any = null;
describe("<App />", () => {
  // runs after all tests have finished
  beforeEach(async () => {
    // setup a DOM element as a render target
    container = document.createElement("div");
    document.body.appendChild(container);
  });
  afterEach(() => {
    // cleanup on exiting
    unmountComponentAtNode(container);
    container.remove();
    container = null;
  });
  //
  it("should renders snapshot", () => {
    const { asFragment } = render(<App />);
    expect(asFragment()).toMatchSnapshot();
  });
  it("should renders component and display loading", () => {
    act(() => {
      render(<App />, container);
    });
    //screen.debug();
    const loadingElement = screen.getByTestId("loading");
    expect(loadingElement.innerHTML).toBe("Loading");
  });

  it("should wait for `loading...` to be removed", async () => {
    await act(async () => {
      render(<App />, container);
    });
    //const loadingElement = screen.getByTestId("loading");
    await waitForElementToBeRemoved(() => screen.queryByText("Loading"));
    //screen.debug();
    const fetchMockCall = jest.spyOn(global, "fetch");
    fetchMockCall.mockImplementationOnce((): any => fetchMock());
  });

  it("displays the robots received from API", async () => {
    const fetchMockCall = jest.spyOn(global, "fetch");
    fetchMockCall.mockImplementationOnce((): any => fetchMock());
    await act(async () => {
      render(<App />, container);
    });
    expect(screen.getByTestId("robotList")).toBeInTheDocument();
    stubbedData.forEach((item) => {
      expect(screen.getByText(item.name)).toBeInTheDocument();
    });
    // remove the mock to ensure tests are completely isolated
    fetchMockCall.mockRestore();
  });

  it("displays error when 404 in the  API", async () => {
    const fetchMockCall = jest.spyOn(global, "fetch");
    fetchMockCall.mockImplementationOnce((): any => rejectFetchMock404());
    await act(async () => {
      render(<App />, container);
    });

    expect(screen.getByTestId("error")).toBeInTheDocument();
    expect(screen.getByTestId("error").innerHTML).toBe("Error 404");

    // remove the mock to ensure tests are completely isolated
    fetchMockCall.mockRestore();
  });

  it("displays something wrong in the API", async () => {
    const fetchMockCall = jest.spyOn(global, "fetch");
    fetchMockCall.mockImplementationOnce((): any => rejectFetchMock());
    //screen.debug();
    await act(async () => {
      render(<App />, container);
    });
    expect(screen.getByTestId("error")).toBeInTheDocument();
    expect(screen.getByTestId("error").innerHTML).toBe("Unknown Error");

    // Remove the mock to ensure tests are completely isolated
    fetchMockCall.mockRestore();
  });

  it("search form enter to filter data", async () => {
    const fetchMockCall = jest.spyOn(global, "fetch");
    fetchMockCall.mockImplementationOnce((): any => fetchMock());
    //screen.debug();
    await act(async () => {
      render(<App />, container);
    });
    userEvent.type(screen.getByTestId("search"), "Naimish");

    expect(screen.getByTestId("robotList").children.length).toBe(1);
    // Remove the mock to ensure tests are completely isolated
    fetchMockCall.mockRestore();
  });

  it("search form with wrong data to find no robots", async () => {
    const fetchMockCall = jest.spyOn(global, "fetch");
    fetchMockCall.mockImplementationOnce((): any => fetchMock());

    await act(async () => {
      render(<App />, container);
    });
    userEvent.type(screen.getByTestId("search"), "ABCD");

    expect(screen.getByTestId("robotList").children.length).toBe(0);
    // Remove the mock to ensure tests are completely isolated
    fetchMockCall.mockRestore();
  });
});
