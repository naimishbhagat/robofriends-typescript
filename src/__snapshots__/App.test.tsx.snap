// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should renders snapshot 1`] = `
<DocumentFragment>
  <h1
    data-testid="loading"
  >
    Loading
  </h1>
</DocumentFragment>
`;
